### PR TITLE
[Dependency Scanning] Specify Source Locations For Missing Module Dependencies

### PR DESCRIPTION
**Pull request by artemcm**
This change modifies the dependency scanner to keep track of source locations of each encountered `import` statement, in order to be able to emit diagnostics with source locations if an import failed to resolve.

- Keep track of each `import` statement's source buffer, line number, and column number when adding it. The dependency scanner utilizes separate compilation instances, and therefore separate Source Managers for scanning `import` statements of user sources and textual interfaces of Swift dependencies. Since import resolution may happen in the main scanner compilation instance while the `import` itself was found by an interface-scanning sub-instance, we cannot simply hold on to the import's `SourceLoc`.
- Add libSwiftScan API for diagnostics to carry above source locations to clients.

Instead of unlocatable:
```
<unknown>0 error: Unable to find module dependency: 'Baz'
```
we will now get rich output:
```
/.../test.swift:2:8: error: Unable to find module dependency: 'Baz'
 1 | import A
 2 | import Baz
   |        |- error: Unable to find module dependency: 'Baz'
   |        `- note: a dependency of main module 'test'
 3 | import C
 4 | import Baz
   |        `- note: also imported here
```

### CODE REVIEW

**Comment by artemcm** [comment]
@swift-ci test

File: test/ScanDependencies/error_source_locations.swift [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+// CHECK-NEXT:   |        `- note: a dependency of main module 'deps'
+// CHECK-NEXT: 8 |
+// CHECK-NEXT: 9 |
+
+// CHECK:      {{.*}}{{/|\\}}Z.swiftinterface:3:8: error: Unable to find module dependency: 'missing_module'
+// CHECK-NEXT: 1 | // swift-interface-format-version: 1.0
+// CHECK-NEXT: 2 | // swift-module-flags: -module-name Z
+// CHECK-NEXT: 3 | import missing_module
+// CHECK-NEXT:   |        |- error: Unable to find module dependency: 'missing_module'
+// CHECK-NEXT:   |        |- note: a dependency of Swift module 'Z': '{{.*}}{{/|\\}}Z.swiftinterface'
```
Comment (artemcm) [inline test/ScanDependencies/error_source_locations.swift:23]: A follow-up change will associate these notes with their corresponding source location of the importee of the given module in the dependency chain.

**Review by nkcsgexi — APPROVED** [review]
Great enhancement on build logs. Thank you!

File: include/swift/DependencyScan/StringUtils.h [hunk 2]
```diff
@@ -13,6 +13,7 @@
 #include "swift-c/CommonString/CommonString.h"
 #include "llvm/ADT/StringRef.h"
 #include "llvm/ADT/StringSet.h"
+#include "llvm/ADT/SmallVector.h"
```
Comment (cachemeifyoucan) [inline include/swift/DependencyScan/StringUtils.h:16]: This is not needed for a header file that is not changed.

File: include/swift-c/DependencyScan/DependencyScan.h [hunk 3]
```diff
@@ -419,9 +422,25 @@ swiftscan_diagnostic_get_message(swiftscan_diagnostic_info_t diagnostic);
 SWIFTSCAN_PUBLIC swiftscan_diagnostic_severity_t
 swiftscan_diagnostic_get_severity(swiftscan_diagnostic_info_t diagnostic);
 
+SWIFTSCAN_PUBLIC swiftscan_source_location_t
```
Comment (cachemeifyoucan) [inline include/swift-c/DependencyScan/DependencyScan.h:425]: Consider version bump?

File: tools/libSwiftScan/libSwiftScan.cpp [hunk 4]
```diff
@@ -698,6 +705,31 @@ swiftscan_diagnostics_set_dispose(swiftscan_diagnostic_set_t* diagnostics){
...
+  return source_location->line_number;
+}
+
+int64_t
+swiftscan_source_location_get_column_number(swiftscan_source_location_t source_location) {
+  return source_location->column_number;
+}
+
+
+void swiftscan_source_location_dispose(swiftscan_source_location_t source_location) {
```
Comment (cachemeifyoucan) [inline tools/libSwiftScan/libSwiftScan.cpp:726]: Do we need this API? The only place can return source location is from a diagnostics and if we call this function, it is going to be double free when dispose diagnostics.

Need good documentation for the ownership for the data structures.

**Review by cachemeifyoucan — COMMENTED** [review]
Nice. We need to send more diagnostics location back (like if a syntax error in the main source module).

File: tools/libSwiftScan/libSwiftScan.cpp (hunk 4 above)
Comment (artemcm) [reply tools/libSwiftScan/libSwiftScan.cpp]: We do not. Removed!

File: include/swift-c/DependencyScan/DependencyScan.h (hunk 3 above)
Comment (artemcm) [reply include/swift-c/DependencyScan/DependencyScan.h:425]: Done. Thanks for the catch.

File: include/swift/DependencyScan/StringUtils.h (hunk 2 above)
Comment (artemcm) [reply include/swift/DependencyScan/StringUtils.h]: Removed.

**Review by owenv — APPROVED** [review]

**Comment by artemcm** [comment]
Waiting on https://github.com/apple/llvm-project/pull/8754 to land and auto-merge before this can land.

**Comment by artemcm** [comment]
@swift-ci test

**Comment by artemcm** [comment]
https://github.com/apple/swift-driver/pull/1614
@swift-ci test

**Comment by artemcm** [comment]
https://github.com/apple/swift-driver/pull/1614
@swift-ci test

**Comment by artemcm** [comment]
https://github.com/apple/swift-driver/pull/1614
@swift-ci test

**Comment by artemcm** [comment]
https://github.com/apple/swift-driver/pull/1614
@swift-ci test

**Comment by artemcm** [comment]
https://github.com/apple/swift-driver/pull/1614
@swift-ci test

**Comment by artemcm** [comment]
https://github.com/apple/swift-driver/pull/1614
@swift-ci test

**Comment by artemcm** [comment]
https://github.com/apple/swift-driver/pull/1614
@swift-ci test

**Comment by artemcm** [comment]
@swift-ci Please Build Toolchain Ubuntu 22.04

**Comment by artemcm** [comment]
https://github.com/apple/swift-driver/pull/1614
@swift-ci test

**Comment by artemcm** [comment]
https://github.com/apple/swift-driver/pull/1614
@swift-ci test

**Pull request merged**